Setup.bat

```diff
@@ -27,14 +36,58 @@ call :MarkStartOfBlock "Setup the git hooks"
 call :MarkEndOfBlock "Setup the git hooks"
 
 call :MarkStartOfBlock "Check dependencies"
-    if not defined UNREAL_HOME (
-        echo Error: Please set UNREAL_HOME environment variable to point to the Unreal Engine folder.
+
+    rem Find the Unreal Engine used to build this project.
+    set UNREAL_ENGINE=""
+
+    rem Get the Unreal Engine used by this project by querying the registry for the engine association found in the .uproject.
+    for /f "delims=" %%A in (' powershell -Command "Get-Childitem -Path %ProjectDirectory% -Recurse -Include *.uproject -File | %% {$_.FullName}" ') do set UPROJECT="%%A"
+    
+    rem If the regex failed then it will return a string containing only a space.
+    if %UPROJECT%==" " (
+        echo Error: Could not find uproject. Please make sure you have passed in the project directory correctly.
         pause
         exit /b 1
     )
 
+    echo Using uproject: %UPROJECT%
+
+    rem Get the Engine Association from the uproject.
+    for /f "delims=" %%A in (' powershell -Command "Select-String -Pattern '..{8}-.{4}-.{4}-.{4}-.{12}.' %UPROJECT% -AllMatches | %% { $_.Matches } | %% { $_.Value }" ') do set ENGINE_ASSOCIATION="%%A"
+    
+    echo Engine association for uproject is: %ENGINE_ASSOCIATION%
+
+    if not %ENGINE_ASSOCIATION%==" " (
+        rem Query the registry for the path to the Unreal Engine using the engine associtation.
+        for /f "usebackq tokens=3*" %%A in (`reg query "HKCU\Software\Epic Games\Unreal Engine\Builds" /v %ENGINE_ASSOCIATION%`) do (
+            set UNREAL_ENGINE="%%A"
+        )
+    )
+
+    rem If there was no engine association then we need to climb the directory path to find the Engine.
+    if %UNREAL_ENGINE%=="" (
+        :climb_parent_directory
+        cd ..
+        if exist Engine (
+            echo Found Engine at: "%cd%"
+            set UNREAL_ENGINE="%cd%"
+        ) else (
+            goto :climb_parent_directory

```

Comment:
Won't it keep looping forever if `Engine` is not found?

Reply:
Tested and you were correct. Added an additional check to exit if we have hit a root directory.